hashintel/hash
# H-1174: remove `tokens` property type in favour of updated `text-content` property type

## 🌟 What is the purpose of this PR?




This PR removes the `tokens` property type from HASH, replacing it with the updated `textual-content` property type which supports both a string and an array of objects as its values.

Previously the hook service was storing the tokens (an array of objects) in a separate `tokens` property, and attaching it to the callout/heading properties. This is problematic with the introduction of entity validation, which requires all properties on an entity to be defined on the entity's type. Making use of the updated `textual-content` property type, which is defined on the block's schemas, addresses this.

It also makes slight adjustments to the page layout:

- decreases the spacing between the blocks and page title
- aligns comments to the start of blocks, not the top of the page

## 🔗 Related links





- H-1174

## 🔍 What does this change?




- updates the `Callout`, `Paragraph` and `Heading` block to make use of the updated `textual-content` property type, and gracefully handle `string` and an array of objects as values

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



## 🐾 Next steps



## 🛡 What tests cover this?



None.

## ❓ How to test this?



1. Checkout the branch locally
2. Clear `hash-workspace-user-blocks-v2` in your local storage, so that the latest blocks are fetched
3. Try using the `Callout`, `Paragraph` and `Heading` blocks in a page, and ensure changes persist (indicating there are no validation errors)

## 📹 Demo



<img width="2056" alt="image" src="https://github.com/hashintel/hash/assets/42802102/93fe4415-c724-4383-a3ea-d17689318528">
